File: backend/app/api/routes/geo.py
from fastapi import APIRouter, HTTPException, Query

from app.services.geo import normalize_lat_lon
from app.services.maps import format_coords, reverse_geocode_details

router = APIRouter(prefix="/geo", tags=["geo"])


@router.get("/reverse")
def reverse_lookup(
    latitude: float = Query(..., ge=-90, le=90),
    longitude: float = Query(..., ge=-180, le=180),
):
    lon, lat = normalize_lat_lon(longitude, latitude)
    details = reverse_geocode_details(lat, lon)
    if not details.location_label and not details.city and not details.pincode:
        raise HTTPException(status_code=404, detail="Could not resolve place name for coordinates")
    return {
        "latitude": lat,
        "longitude": lon,
        "location_label": details.location_label,
        "city": details.city,
        "state": details.state,
        "pincode": details.pincode,
        "coords_label": format_coords(lat, lon),
    }

File: frontend/src/services/geo.ts
import { api } from "./api";

export type ReverseGeoResult = {
  latitude: number;
  longitude: number;
  location_label: string | null;
  city: string | null;
  state: string | null;
  pincode: string | null;
  coords_label?: string | null;
};

const PLACEHOLDER_LOCATION_LABELS = new Set([
  "detected from device",
  "current location",
]);

/** True when a location label is worth showing in the UI. */
export function isMeaningfulLocationLabel(label?: string | null): boolean {
  const text = label?.trim();
  if (!text) return false;
  return !PLACEHOLDER_LOCATION_LABELS.has(text.toLowerCase());
}

/** Resolve coordinates to local place name, city, and pincode via the backend. */
export async function reverseGeocodeDetails(
  latitude: number,
  longitude: number,
): Promise<ReverseGeoResult | null> {
  try {
    const { data } = await api.get<ReverseGeoResult>("/geo/reverse", {
      params: { latitude, longitude },
      timeout: 10000,
    });
    return data;
  } catch {
    return null;
  }
}

/** Resolve coordinates to a short place name via the backend. */
export async function reverseGeocode(
  latitude: number,
  longitude: number,
): Promise<string | null> {
  const details = await reverseGeocodeDetails(latitude, longitude);
  return details?.location_label?.trim() || details?.city?.trim() || null;
}
